hedera-node/src/main/java/com/hedera/services/store/schedule/HederaScheduleStore.java

```diff
@@ -183,49 +251,245 @@ public boolean isCreationPending() {
+
+		var equalityKey = new ScheduleEqualityVirtualKey(schedule.equalityCheckKey());
+		var byEquality = schedules.get().byEquality().get(equalityKey);
+		if (byEquality != null) {
+			var existingId = byEquality.getIds().get(schedule.equalityCheckValue());
+
+			if (existingId != null) {
+
+				final var extantId = EntityNum.fromLong(existingId).toGrpcScheduleId();
+
+				if (exists(extantId)) {
+					return Pair.of(extantId, get(extantId));
+				}
+			}
 		}
 
 		return Pair.of(null, schedule);
 	}
 
+	@Override
+	public ResponseCodeEnum preMarkAsExecuted(ScheduleID id) {
+		var status = usabilityCheck(id, false, null);
+		if (status != OK) {
+			return status;
+		}
+		return OK;
+	}
+
 	@Override
 	public ResponseCodeEnum markAsExecuted(ScheduleID id, Instant consensusTime) {
-		var status = usabilityCheck(id, false);
+		var status = usabilityCheck(id, false, null);
 		if (status != OK) {
 			return status;
 		}
-		final var offset = properties.triggerTxnWindBackNanos();
-		final var executionTime = consensusTime.plusNanos(offset);
-		apply(id, schedule -> schedule.markExecuted(executionTime));
+		apply(id, schedule -> schedule.markExecuted(consensusTime));
 		return OK;
 	}
 
 	@Override
-	public void expire(EntityId entityId) {
-		var id = entityId.toGrpcScheduleId();
+	public void expire(ScheduleID id) {
 		if (id.equals(pendingId)) {
 			throw new IllegalArgumentException(String.format(
 					"Argument 'id=%s' refers to a pending creation!",
 					readableId(id)));
 		}
-		var schedule = get(id);
-		schedules.get().remove(fromLong(entityId.num()));
-		extantSchedules.remove(schedule);
+
+		throwIfMissing(id);
+
+
+		var idToDelete = new EntityNumVirtualKey(fromScheduleId(id));
+		var existingSchedule = schedules.get().byId().remove(idToDelete);
+
+		if (existingSchedule != null) {
+
+			var secondKey = new SecondSinceEpocVirtualKey(existingSchedule.calculatedExpirationTime().getSeconds());
+
+			var bySecond = schedules.get().byExpirationSecond().get(secondKey);
+
+			if (bySecond != null) {
+				bySecond = bySecond.asWritable();
+				bySecond.removeId(existingSchedule.calculatedExpirationTime(), idToDelete.getKeyAsLong());
+
+				if (bySecond.getIds().isEmpty()) {
+					schedules.get().byExpirationSecond().remove(secondKey);
+				} else {
+					schedules.get().byExpirationSecond().put(secondKey, bySecond);
+				}
+			}
+
+
+			var equalityKey = new ScheduleEqualityVirtualKey(existingSchedule.equalityCheckKey());
+
+			var byEquality = schedules.get().byEquality().get(equalityKey);
+
+			if (byEquality != null) {
+				byEquality = byEquality.asWritable();
+				byEquality.remove(existingSchedule.equalityCheckValue(), idToDelete.getKeyAsLong());
+
+				if (byEquality.getIds().isEmpty()) {
+					schedules.get().byEquality().remove(equalityKey);
+				} else {
+					schedules.get().byEquality().put(equalityKey, byEquality);
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean advanceCurrentMinSecond(Instant consensusTime) {
```

Comment:
Why is this method public? It seems more appropriate as a `private` method that is called at the beginning of `nextSchedulesToExpire()`.